File: app/src/main/java/nic/goi/aarogyasetu/utility/LocationUtils.java
/*
 * Copyright 2020 Government of India
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package nic.goi.aarogyasetu.utility;

import android.app.Activity;
import android.content.Context;
import android.content.IntentSender;
import android.location.LocationManager;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.google.android.gms.common.api.ResolvableApiException;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResponse;
import com.google.android.gms.location.SettingsClient;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;

import nic.goi.aarogyasetu.R;

public class LocationUtils {
    public static final int LOCATION_REQUEST = 1245;
    private Context context;

    public LocationUtils(Context context) {
        this.context = context;
    }

    public void turnLocationOn(TurnLocationListener turnLocationListener) {
        LocationRequest locationRequest = LocationRequest.create();
        locationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);

        LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder()
                .addLocationRequest(locationRequest);
        builder.setAlwaysShow(true);
        LocationSettingsRequest mLocationSettingsRequest = builder.build();

        SettingsClient client = LocationServices.getSettingsClient(context);
        Task<LocationSettingsResponse> task = client.checkLocationSettings(mLocationSettingsRequest);

        task.addOnSuccessListener((Activity) context, locationSettingsResponse -> {

            if (turnLocationListener != null) {
                turnLocationListener.locationStatus(true);
            }
        });

        task.addOnFailureListener((Activity) context, new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                if (e instanceof ResolvableApiException) {
                    // Location settings are not satisfied, but this can be fixed
                    // by showing the user a dialog.
                    try {
                        // Show the dialog by calling startResolutionForResult(),
                        // and check the result in onActivityResult().
                        ResolvableApiException resolvable = (ResolvableApiException) e;
                        resolvable.startResolutionForResult(((Activity) context),
                                LOCATION_REQUEST);
                    } catch (IntentSender.SendIntentException sendEx) {
                        // Ignore the error.
                        Toast.makeText(context, LocalizationUtil.getLocalisedString(context, R.string.error_location), Toast.LENGTH_LONG).show();
                    }
                }
            }
        });
    }

    public interface TurnLocationListener {
        void locationStatus(boolean isTurnOn);
    }
}

File: app/src/main/java/nic/goi/aarogyasetu/utility/LocationUtils.kt
package nic.goi.aarogyasetu.utility

import android.app.Activity
import android.content.Context
import android.content.IntentSender
import android.location.LocationManager
import android.widget.Toast

import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.LocationSettingsRequest
import com.google.android.gms.location.LocationSettingsResponse
import com.google.android.gms.location.SettingsClient
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.Task

import nic.goi.aarogyasetu.R

class LocationUtils(private val context: Context) {

    fun turnLocationOn(turnLocationListener: TurnLocationListener?) {
        val locationRequest = LocationRequest.create()
        locationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY)

        val builder = LocationSettingsRequest.Builder()
            .addLocationRequest(locationRequest)
        builder.setAlwaysShow(true)
        val mLocationSettingsRequest = builder.build()

        val client = LocationServices.getSettingsClient(context)
        val task = client.checkLocationSettings(mLocationSettingsRequest)

        task.addOnSuccessListener(context as Activity, { locationSettingsResponse ->

            turnLocationListener?.locationStatus(true)
        })

        task.addOnFailureListener(context as Activity, object : OnFailureListener() {
            @Override
            fun onFailure(@NonNull e: Exception) {
                if (e is ResolvableApiException) {
                    // Location settings are not satisfied, but this can be fixed
                    // by showing the user a dialog.
                    try {
                        // Show the dialog by calling startResolutionForResult(),
                        // and check the result in onActivityResult().
                        val resolvable = e as ResolvableApiException
                        resolvable.startResolutionForResult(
                            context as Activity,
                            LOCATION_REQUEST
                        )
                    } catch (sendEx: IntentSender.SendIntentException) {
                        // Ignore the error.
                        Toast.makeText(
                            context,
                            LocalizationUtil.getLocalisedString(context, R.string.error_location),
                            Toast.LENGTH_LONG
                        ).show()
                    }

                }
            }
        })
    }

    interface TurnLocationListener {
        fun locationStatus(isTurnOn: Boolean)
    }

    companion object {
        val LOCATION_REQUEST = 1245
    }
}
